### ISSUE TITLE
FactorRange regression: BokehJS crashes if data has values not in FactorRange factors.

### ISSUE THREAD

**Issue by lukauskas** [bokeh/bokeh#7084]
This is probably related to  #7083 but at the same time sufficiently different, hence new issue number.

#### ALL software version info (bokeh, python, notebook, OS, browser, any other relevant packages)

Bokeh (crashing) 0.12.10, 0.12.7
Bokeh (not crashing): 0.12.6

OS: docker image [lukauskas/bokeh](https://hub.docker.com/r/lukauskas/bokeh/)
Python: 3.6

Browser: Chrome 61.0.3163.100

#### Description of expected behavior and the observed behavior

Since 0.12.7 onwards, bokehjs would crash if asked to plot something on a factor axis, when given a value that is not in `factors`.
Specifically, in the example below I create a figure with a FactorRange having factors `'a'`, `'b'` and `'c'`.

I then proceed to draw a circle from `ColumnDataSource`. Now the tricky bit is that my ColumnSource also has an item at the location `x='d'`, `y=2`. 

Now I am not sure what the expected behaviour should be, but is definitely not to crash silently.
The options are:

1) Display the plot without the data that is not in factors (this would be my preferred option, also how it worked in 0.12.6). When combined with  #7083 this would allow to 'zoom in' into factor ranges without changing the source.

or, at least:

2) Throw an useful error message so it is easier to find what the problem is.

#### Complete, minimal, self-contained example code that reproduces the issue

```
from bokeh.models import FactorRange, Button, ColumnDataSource
from bokeh.plotting import figure, curdoc
from bokeh.layouts import layout

button = Button()
x_range = FactorRange(factors=['a', 'b', 'c'])

source = ColumnDataSource({'x': ['a', 'd'], 'y': [1, 2]})

fig = figure(x_range=x_range)
fig.circle(x='x', y='y', size=20, color='navy', source=source)

l = layout([[fig]])
curdoc().add_root(l)
```

To reproduce it either run the code directly via `bokeh serve`, or place it `app/main.py` and use the aforementioned docker image with this `Dockerfile`:

```
ARG VERSION=0.12.10
FROM lukauskas/bokeh:${VERSION}
COPY ./app /app
```

```
docker build --build-arg VERSION="0.12.10" -t bokeh .
docker run --rm -p 5100:5100 --name bokeh_cont bokeh
```

#### browser JavaScript console output (bokeh=0.12.10)
```
Error rendering Bokeh items  TypeError: Cannot read property 'value' of undefined
    at e.r.FactorRange.e._lookup (factor_range.js:288)
    at e.r.FactorRange.e.synthetic (factor_range.js:223)
    at e.<anonymous> (factor_range.js:239)
    at e.r.FactorRange.e.v_synthetic (factor_range.js:242)
    at e.r.GlyphView.r.set_data (glyph.js:301)
    at e.r.GlyphRendererView.e.set_data (glyph_renderer.js:160)
    at e.r.GlyphRendererView.e.initialize (glyph_renderer.js:87)
    at e.t (view.js:42)
    at e [as constructor] (dom_view.js:21)
    at e [as constructor] (renderer.js:25)
```

In short `this._mapping['d']` returns `undefined` (as it is not in the factors list), which then does not have a `value` property and hence the `TypeError`.

**Comment by bryevdv** [comment]
What's the right thing to do here? Just silently ignore values that are not in the range? I can't really think of any alternative.

**Comment by philippjfr** [comment]
>Just silently ignore values that are not in the range?

That's what it used to do before the categorical refactor, so I think it's the correct (or at least most consistent) thing to do.
